banyand/stream/filter.go

```diff
@@ -0,0 +1,352 @@
+						valueType: valueType,
+					}
+				}
+			}
+			if fr.orderByTimestampDesc() {
+				fr.data[i].idx = len(fr.data[i].timestamps) - 1
+			}
+			fr.data = append(fr.data, fr.data[i])
+		}
+		fr.data = fr.data[n:]
+		fr.loaded = true
+		heap.Init(fr)
+	}
+	if len(fr.data) == 0 {
+		return nil
+	}
+	if len(fr.data) == 1 {
+		r := &pbv1.StreamResult{}
+		bc := fr.data[0]
+		bc.copyAllTo(r, fr.orderByTimestampDesc())
+		fr.data = fr.data[:0]
+		return r
+	}
+	return fr.merge()
+}
+
+func (fr *filterResult) Release() {
+	for i, v := range fr.data {
+		releaseBlockCursor(v)
+		fr.data[i] = nil
+	}
+	fr.data = fr.data[:0]
+	for i := range fr.snapshots {
+		fr.snapshots[i].decRef()
+	}
+	fr.snapshots = fr.snapshots[:0]
+}
+
+func (fr filterResult) Len() int {
+	return len(fr.data)
+}
+
+func (fr filterResult) Less(i, j int) bool {
+	leftTS := fr.data[i].timestamps[fr.data[i].idx]
+	rightTS := fr.data[j].timestamps[fr.data[j].idx]
+	if fr.orderByTS {
+		if fr.ascTS {
+			return leftTS < rightTS
+		}
+		return leftTS > rightTS
+	}
+	leftSIDIndex := fr.sidToIndex[fr.data[i].bm.seriesID]
+	rightSIDIndex := fr.sidToIndex[fr.data[j].bm.seriesID]
+	return leftSIDIndex < rightSIDIndex
+}
+
+func (fr filterResult) Swap(i, j int) {
+	fr.data[i], fr.data[j] = fr.data[j], fr.data[i]
+}
+
+func (fr *filterResult) Push(x interface{}) {
+	fr.data = append(fr.data, x.(*blockCursor))
+}
+
+func (fr *filterResult) Pop() interface{} {
+	old := fr.data
+	n := len(old)
+	x := old[n-1]
+	fr.data = old[0 : n-1]
+	return x
+}
+
+func (fr *filterResult) orderByTimestampDesc() bool {
+	return fr.orderByTS && !fr.ascTS
+}
+
+func (fr *filterResult) merge() *pbv1.StreamResult {
+	step := 1
+	if fr.orderByTimestampDesc() {
+		step = -1
+	}
+	result := &pbv1.StreamResult{}
+	var lastSid common.SeriesID
+
+	for fr.Len() > 0 {
+		topBC := fr.data[0]
+		if lastSid != 0 && topBC.bm.seriesID != lastSid {
+			return result
+		}
+		lastSid = topBC.bm.seriesID
+
+		topBC.copyTo(result)
+		topBC.idx += step
+
+		if fr.orderByTimestampDesc() {
+			if topBC.idx < 0 {
+				heap.Pop(fr)
+			} else {
+				heap.Fix(fr, 0)
+			}
+		} else {
+			if topBC.idx >= len(topBC.timestamps) {
+				heap.Pop(fr)
+			} else {
+				heap.Fix(fr, 0)
+			}
+		}
+	}
+
+	return result
+}
+
+func (s *stream) Filter(ctx context.Context, sfo pbv1.StreamFilterOptions) (srp pbv1.StreamResultPuller, err error) {
+	if sfo.TimeRange == nil || len(sfo.Entities) < 1 {
+		return nil, errors.New("invalid query options: timeRange and series are required")
+	}
+	if len(sfo.TagProjection) == 0 {
+		return nil, errors.New("invalid query options: tagProjection is required")
+	}
+	db := s.databaseSupplier.SupplyTSDB()
+	var result filterResult
+	if db == nil {
+		return srp, nil
+	}
+	tsdb := db.(storage.TSDB[*tsTable, option])
+	tabWrappers := tsdb.SelectTSTables(*sfo.TimeRange)
+	sort.Slice(tabWrappers, func(i, j int) bool {
+		return tabWrappers[i].GetTimeRange().Start.Before(tabWrappers[j].GetTimeRange().Start)
+	})
```

Comment:
Why did you sort the tables here? The table list is already in ascending order by default.